### PR TITLE
Rule Safe Delete Refactoring #13 [MK10]

### PR DESCRIPTION
**Pull request by Valvon**
Adds safe delete refactoring for rules and rule files.
Only works from 2023.3 and up.
Minimal Build version needs to be lifted.

### CODE REVIEW

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    override fun findConflicts(
+        element: PsiElement,
+        allElementsToDelete: Array<out PsiElement>,
+    ): MutableCollection<String> {
+        /*
+         * Explicitly check for conflicts in the context of the tag, its parent and the $element to be deleted
+         * */
+        val messages = mutableListOf<String>()
+        val possibleConflicts =
+            allElementsToDelete.filter { (it as? XmlTag?)?.name == "requiresConcept" }.map { it as XmlTag }
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt:112]:
```suggestion
            allElementsToDelete.filterIsInstance<XmlTag>().filter { it.name == "requiresConcept" }
```
Nit: writing it this way wsould be more idiomatic, you can also take a look at [mapNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/map-not-null.html), [filterNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/filter-not-null.html) and [filterIsInstance](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/filter-is-instance.html)

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        val noDeletableParents = possibleConflicts.filter { !allElementsToDelete.contains(it.parent) }
+        val relatedConflicts =
+            noDeletableParents.filter { (element as? XmlTag)?.getAttributeValue("id") == it.getAttributeValue("refId") }
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt:115]: Nit: Would this expression not also evaluate to true if both sides were `null`?

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+     * be excluded.
+     *
+     * @param project the project where the refactoring happens.
+     * @param usages all usages to be processed by the refactoring.
+     * @return the filtered list of usages, or null if the user has cancelled the refactoring.
+     */
+    override fun preprocessUsages(
+        project: Project,
+        usages: Array<out UsageInfo>,
+    ): Array<UsageInfo> = usages.map2Array { it }
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt:145]: Nit: would [toArray()](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/-abstract-collection/to-array.html) not be enough here?

File: src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        val manager = DomManager.getDomManager(element?.project)
+
+        val parentTag = element?.parent?.parent?.parent // attribute -> value -> whitespace -> tag
+        val parentXmlTag = parentTag as? XmlTag?
+        val domParentTag = manager.getDomElement(parentXmlTag)
+
+        val tag = element as? XmlTag?
+        val domElement = manager.getDomElement(tag)
+
+        return domElement is RuleBase || domParentTag is RuleBase
```
Comment (A-K-O-R-A) [inline src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt:202]: Nit: add a comment explaining why the parent also gets checked

**Review by A-K-O-R-A — CHANGES_REQUESTED** [review]

File: gradle.properties
[the remark this answers is not in the document]
Comment (Valvon) [reply gradle.properties:1]: removed it

**Review by A-K-O-R-A — APPROVED** [review]

**Pull request merged**